pkg/controller/completed_workflows.go

```diff
@@ -73,10 +75,36 @@ func CalculateHoursToDelete(retentionPeriodHours, currentHourOfDay int) []string
 	return hoursToDelete
 }
 
+//Calculates a list of all the hours that should be kept given the current time and the retentionperiod in hours
+func CalculateHoursToKeep(retentionPeriodHours int, currentTime time.Time) []string {
+	hoursToKeep := make([]string, 0, retentionPeriodHours+1)
+	for i := 0; i <= retentionPeriodHours; i++ {
+		hoursToKeep = append(hoursToKeep, SplitTimeToMeetsLabelFormat(currentTime))
+		currentTime = currentTime.Add(-1 * time.Hour)
+	}
+	return hoursToKeep
+}
+
 // Creates a new selector that selects all completed workflows and workflows with completed hour label outside of the
 // retention window
 func CompletedWorkflowsSelectorOutsideRetentionPeriod(retentionPeriodHours int, currentTime time.Time) *v1.LabelSelector {
-	hoursToDelete := CalculateHoursToDelete(retentionPeriodHours, currentTime.Hour())
+	hoursToKeep := CalculateHoursToKeep(retentionPeriodHours, currentTime)
+	s := CompletedWorkflowsLabelSelector()
+	s.MatchExpressions = append(s.MatchExpressions, v1.LabelSelectorRequirement{
+		Key:      completedTimeKey,
+		Operator: v1.LabelSelectorOpNotIn,
+		Values:   hoursToKeep,
+	})
+	s.MatchExpressions = append(s.MatchExpressions, v1.LabelSelectorRequirement{
```

Comment:
It seems this additional condition would result not matching anything because new resources have no `hour-of-day` label.